google/generativeai/caching.py

```diff
@@ -14,90 +14,124 @@
 # limitations under the License.
 from __future__ import annotations
 
+import copy
 import dataclasses
-import datetime
 from typing import Any, Iterable, Optional
 
 from google.generativeai import protos
-from google.generativeai.types.model_types import idecode_time
 from google.generativeai.types import caching_types
 from google.generativeai.types import content_types
+from google.generativeai import string_utils
 from google.generativeai.utils import flatten_update_paths
 from google.generativeai.client import get_default_cache_client
 
 from google.protobuf import field_mask_pb2
-import google.ai.generativelanguage as glm
 
 
-@dataclasses.dataclass
+@string_utils.prettyprint
 class CachedContent:
     """Cached content resource."""
 
-    name: str
-    model: str
-    create_time: datetime.datetime
-    update_time: datetime.datetime
-    expire_time: datetime.datetime
+    def __init__(self, name):
+        """Fetches a `CachedContent` resource.
 
-    # NOTE: Automatic CachedContent deletion using contextmanager is not P0(P1+).
-    # Adding basic support for now.
-    def __enter__(self):
-        return self
+        Identical to `CachedContent.get`.
+
+        Args:
+            name: The resource name referring to the cached content.
+        """
+        client = get_default_cache_client()
 
-    def __exit__(self, exc_type, exc_value, exc_tb):
-        self.delete()
+        if "cachedContents/" not in name:
+            name = "cachedContents/" + name
 
-    def _to_dict(self) -> protos.CachedContent:
+        request = protos.GetCachedContentRequest(name=name)
+        response = client.get_cached_content(request)
+        self._proto = response
+
+    @property
+    def name(self) -> str:
+        return self._proto.name
+
+    @property
+    def model(self) -> str:
+        return self._proto.model
+
+    @property
+    def display_name(self) -> str:
+        return self._proto.display_name
+
+    @property
+    def usage_metadata(self) -> protos.CachedContent.UsageMetadata:
+        return self._proto.usage_metadata
+
+    @property
+    def create_time(self) -> datetime.datetime:
+        return self._proto.create_time
+    @property
+    def update_time(self) -> datetime.datetime:
+        return self._proto.update_time
+    @property
+    def expire_time(self) -> datetime.datetime:
+        return self._proto.expire_time
+
+    def _update(self, updates):
+        if not hasattr(self, "_proto"):
+            self._proto = protos.CachedContent(updates)
+            return
+
+        if not isinstance(updates, CachedContent):
+            updates = protos.CachedContent(updates)
+        updates = type(updates).to_dict(updates, including_default_value_fields=False)
+        for key, value in updates.items():
+            setattr(self._proto, key, value)
```

Comment:
Current implementation is a little hard to read. I think this would be more approachable. What do you think, @MarkDaoust ?

```suggestion
     def __new__(cls, updates=None, **kwargs):
         self = super().__new__(cls)
         
         if not updates:
             return self
         
         if not hasattr(self, "_proto"):
            self._proto = protos.CachedContent(updates)
            return

          if not isinstance(updates, CachedContent):
              updates = protos.CachedContent(updates)
          updates = type(updates).to_dict(updates, including_default_value_fields=False)
          for key, value in updates.items():
              setattr(self._proto, key, value)
           
           return self
```

Then replace these lines:
result = object.__new__(CachedContent)
cc = response._update(response)

with:
cc = cls.__new__(response)


calling `object`'s  __new__ is a little weird, we are doing the same thing essentially from super().__new__() but it is inside the __new__. What say?

Reply:
I ended up putting this in a 

```
@classmethod
def _from_obj(cls, obj)
```